### PR TITLE
Deprecate `boxplot(..., whis="range")`.

### PR DESCRIPTION
**Pull request by anntzer**
It's a special-case that needs its own testing and documentation, when
it can be perfectly well folded into `whis=(percentile, percentile)`
(specifically, `whis=(0, 100)`).

Noted while reviewing #14358.

attn @phobson in can you think it's really worth keeping the special-cased API.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by phobson** [comment]
(BTW, I can't parse the CircleCI failure)

**Comment by anntzer** [comment]
looks like an unrelated failure to resolve references to dateutil in the docs due to problems with their online docs.

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -3503,18 +3503,23 @@ def boxplot(self, x, notch=None, sym=None, vert=None, whis=None,
             everything is drawn horizontally.
 
         whis : float, sequence, or string (default = 1.5)
```
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py]:
```suggestion
         whis : float or (float, float) or str, default = 1.5
```

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -3503,18 +3503,23 @@ def boxplot(self, x, notch=None, sym=None, vert=None, whis=None,
...
-            whiskers to the min and max of the data.
+            The position of the whiskers.
+
+            If a float, the lower whisker is at position ``Q1 - whis*(Q3-Q1)``,
+            and the upper whisker at position ``Q3 + whis*(Q3-Q1)``, where Q1
+            and Q3 are the first and third quartiles.
+
+            If a pair of floats, they indicate the percentiles at which to
+            draw the whiskers (e.g., (5, 95)).  In particular, setting this to
+            [0, 100] results in whiskers covering the whole range of the data.
```
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py]:
```suggestion
            (0, 100) results in whiskers covering the whole range of the data.
```
Keep this consistent and not switch between tuples and lists. While both are possible, tuple seems more appropriate semantically.

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -3503,18 +3503,23 @@ def boxplot(self, x, notch=None, sym=None, vert=None, whis=None,
...
+            "range" is a deprecated synonym for (0, 100).
+
+            In the edge case where ``Q1 == Q3``, *whis* is automatically set
+            to [0, 100] (cover the whole range of the data) if *autorange* is
```
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py]:
```suggestion
            to (0, 100) (cover the whole range of the data) if *autorange* is
```

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -3503,18 +3503,23 @@ def boxplot(self, x, notch=None, sym=None, vert=None, whis=None,
...
-            Beyond the whiskers, data
-            are considered outliers and are plotted as individual
-            points. Alternatively, set
-            this to an ascending sequence of percentile (e.g., [5, 95])
-            to set the whiskers at specific percentiles of the data.
-            Finally, ``whis`` can be the string ``'range'`` to force the
-            whiskers to the min and max of the data.
+            The position of the whiskers.
+
+            If a float, the lower whisker is at position ``Q1 - whis*(Q3-Q1)``,
```
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py]: To be precise,  the whisker extends at most that far, but ends at the last data point within this range

> the lowest datum still within 1.5 IQR of the lower quartile, and the highest datum still within 1.5 IQR of the upper quartile (often called the Tukey boxplot)

https://en.wikipedia.org/wiki/Box_plot#Types (second bullet point).

File: lib/matplotlib/cbook/__init__.py [hunk 5]
```diff
@@ -1039,20 +1039,23 @@ def boxplot_stats(X, whis=1.5, bootstrap=None, labels=None,
...
-        ascending sequence of percentiles (e.g., [5, 95]) to set the
-        whiskers at specific percentiles of the data. Finally, `whis`
-        can be the string ``'range'`` to force the whiskers to the
-        minimum and maximum of the data. In the edge case that the 25th
-        and 75th percentiles are equivalent, `whis` can be automatically
-        set to ``'range'`` via the `autorange` option.
+    whis : float or (float, float) (default = 1.5)
+        The position of the whiskers.
+
+        If a float, the lower whisker is at position ``Q1 - whis*(Q3-Q1)``, and
```
Comment (timhoffm) [inline lib/matplotlib/cbook/__init__.py]: All comments from above apply.

**Comment by anntzer** [comment]
thanks, all comments handled.

**Comment by timhoffm** [comment]
+1 on the doc change +0.5 on the deprecation. Supporting four strings (three of which seem completely undocumented) as alternative to `[0, 100]` is insane. Generally, I would argue that  a string is simpler to understand compared to `whis=[0, 100]`, but I wouldn't know which one of these to keep, they all feel not quite right, too. So let's just remove them.

**Pull request merged**